File: 0945-snakes-and-ladders/0945-snakes-and-ladders.java
class Solution {
    public int snakesAndLadders(int[][] board) {
        int n = board.length;
        int max = n * n;
        int[] boardLin = makeBoardLineer(board);
        boolean[] isVisited = new boolean[max];
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(0);
        isVisited[0] = true;

        int steps = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            while (size-- > 0) {
                int curr = queue.poll();

                if (curr == max - 1) return steps;

                for (int move = 1; move <= 6; move++) {
                    int next = curr + move;
                    if (next >= max) continue;

                    int dest = (boardLin[next] != -1) ? boardLin[next] - 1 : next;
                    if (!isVisited[dest]) {
                        isVisited[dest] = true;
                        queue.offer(dest);
                    }
                }
            }
            steps++;
        }
        return -1;
    }

    public static int[] makeBoardLineer(int[][] board) {
        int n = board.length;
        int[] res = new int[n * n];
        int index = 0;
        boolean leftToRight = true;

        for (int i = n - 1; i >= 0; i--) {
            if (leftToRight) {
                for (int j = 0; j < n; j++) {
                    res[index++] = board[i][j];
                }
            } else {
                for (int j = n - 1; j >= 0; j--) {
                    res[index++] = board[i][j];
                }
            }
            leftToRight = !leftToRight;
        }

        return res;
    }
}

File: 0945-snakes-and-ladders/0945-snakes-and-ladders.cpp
class Solution {
public:
    int snakesAndLadders(vector<vector<int>>& board) {
        int n = board.size(), res = 1;
        int* arr = boustrophedonToArray(board);
        vector<bool> isVisited(n*n, false);
        queue<int> curr, next;

        curr.push(0);
        isVisited[0] = true;
        while (!curr.empty()) {
            int x = curr.front();
            curr.pop();
            for (int i = 1; i < min(7, n*n - x); i++) {
                int dest = arr[x + i] == -1 ? x + i: arr[x + i] - 1;
                if (dest == n*n - 1) {
                    delete[] arr;
                    return res;
                }
                if (isVisited[dest]) continue;
                
                isVisited[dest] = true;
                next.push(dest);
            }
            if (curr.empty()) {
                res++;
                curr = move(next);
            }
        }
        delete[] arr;
        return -1;
    }

    int* boustrophedonToArray(vector<vector<int>>& board) {
        int n = board.size(), index = 0;
        int* arr = new int[n*n];
        bool leftToRight = true;
        for (int i = n - 1; i >= 0; i--) {
            if (leftToRight) {
                for (int j = 0; j < n; j++) {
                    arr[index] = board[i][j];
                    index++;
                }  
            } else {
                for (int j = n - 1; j >= 0; j--) {
                    arr[index] = board[i][j];
                    index++;
                } 
            }
            leftToRight = !leftToRight;
        }
        return arr;
    }
};